packages/analytics-js/src/components/eventRepository/EventRepository.ts

```diff
@@ -1,43 +1,92 @@
+import { state } from '@rudderstack/analytics-js/state';
 import { IEventRepository } from './types';
 import { IPluginsManager } from '../pluginsManager/types';
 import { RudderEvent } from '../eventManager/types';
-import { DEFAULT_QUEUE_OPTIONS } from './constants';
 import { defaultPluginsManager } from '../pluginsManager';
+import {
+  DATA_PLANE_QUEUE_EXT_POINT_PREFIX,
+  DESTINATIONS_QUEUE_EXT_POINT_PREFIX,
+} from './constants';
 
+/**
+ * Event repository class responsible for queuing events for further processing and delivery
+ */
 class EventRepository implements IEventRepository {
-  dataPlaneEventsQueue: Queue;
-  deviceEventsQueue: Queue;
+  errorHandler?: IErrorHandler;
+  logger?: ILogger;
   pluginsManager: IPluginsManager;
 
-  constructor(queueOptions: QueueOptions, pluginsManager: IPluginsManager) {
+  /**
+   *
+   * @param pluginsManager Plugins manager instance
+   * @param errorHandler Error handler object
+   * @param logger Logger object
+   */
+  constructor(pluginsManager: IPluginsManager, errorHandler?: IErrorHandler, logger?: ILogger) {
     this.pluginsManager = pluginsManager;
+    this.errorHandler = errorHandler;
+    this.logger = logger;
+    this.onError = this.onError.bind(this);
 
-    // Override default options with user provided options
-    const dpQueueOptions = mergeDeepRight(DEFAULT_QUEUE_OPTIONS, queueOptions);
-
-    this.dataPlaneEventsQueue = new Queue('rudder', dpQueueOptions, (item, done) => {
-      pluginsManager.invoke('dataplaneEvent.process', item, done);
-    });
-
-    this.deviceEventsQueue = new Queue('rs_dm', DEFAULT_QUEUE_OPTIONS, (item, done) => {
-      pluginsManager.invoke('deviceEvent.process', item, done);
-    });
+    this.pluginsManager.invoke(
+      `${DATA_PLANE_QUEUE_EXT_POINT_PREFIX}.init`,
+      state.lifecycle.writeKey,
+      state.lifecycle.activeDataplaneUrl,
+      state.loadOptions.value.queueOptions,
+    );
+    this.pluginsManager.invoke(
+      `${DESTINATIONS_QUEUE_EXT_POINT_PREFIX}.init`,
+      state.loadOptions.value.destinationsQueueOptions,
+    );
   }
 
+  /**
+   * Initializes the event repository
+   */
   init(): void {
-    this.dataPlaneEventsQueue.start();
-    this.deviceEventsQueue.start();
+    this.pluginsManager.invoke(`${DATA_PLANE_QUEUE_EXT_POINT_PREFIX}.start`);
+    this.pluginsManager.invoke(`${DESTINATIONS_QUEUE_EXT_POINT_PREFIX}.start`);
   }
 
+  /**
+   * Enqueues the event for processing
+   * @param event RudderEvent object
+   * @param callback API callback function
+   */
   enqueue(event: RudderEvent, callback?: ApiCallback): void {
-    this.dataPlaneEventsQueue.addItem({ event, callback });
-    this.deviceEventsQueue.addItem({ event, callback });
+    this.pluginsManager.invoke(`${DATA_PLANE_QUEUE_EXT_POINT_PREFIX}.enqueue`, event, this.logger);
+    this.pluginsManager.invoke(`${DESTINATIONS_QUEUE_EXT_POINT_PREFIX}.enqueue`, event);
+
+    // Invoke the callback if it exists
+    try {
+      callback?.(event);
```

Comment:
should it not be that callback gets called after the event has been sent and server replies with 200 ok?

Reply:
Ideally, yes. But this is the current behaviour with v1.1 as well. I'll add this item into our improvements doc.

Reply:
I believe, that work will be done on dataplane queue